src/libinjector/linux/linux_utils.c

```diff
@@ -0,0 +1,206 @@
+#include "linux_utils.h"
+#include "linux_debug.h"
+#include <sys/mman.h>
+#include <fcntl.h>
+
+addr_t find_vdso(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    addr_t process_base = drakvuf_get_current_process(drakvuf, info);
+    PRINT_DEBUG("Process base: %lx\n", process_base);
+
+    ACCESS_CONTEXT(ctx,
+        .translate_mechanism = VMI_TM_PROCESS_PID,
+        .pid = info->proc_data.pid
+    );
+
+    addr_t addr = 0;
+    size_t offset = 0;
+
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+
+    // task_struct to mm
+    if (!drakvuf_get_kernel_struct_member_rva(drakvuf, "task_struct", "mm", &offset))
+    {
+        PRINT_DEBUG("Failed to get mm offset\n");
+        drakvuf_release_vmi(drakvuf);
+        return 0;
+    }
+
+    PRINT_DEBUG("mm offset: %ld\n", offset);
+    ctx.addr = process_base + offset;
+
+    // since mm is a pointer
+    if (VMI_SUCCESS != vmi_read_64(vmi, &ctx, &addr))
+    {
+        PRINT_DEBUG("Failed to read mm address\n");
+        drakvuf_release_vmi(drakvuf);
+        return 0;
+    }
+
+    PRINT_DEBUG("Got mm address: %lx\n", addr);
+
+    // mm_struct to vdso ( it will directly parse the anonymous structure of context in between )
+    if (!drakvuf_get_kernel_struct_member_rva(drakvuf, "mm_struct", "vdso", &offset))
+    {
+        PRINT_DEBUG("Failed to get vdso offset\n");
+        drakvuf_release_vmi(drakvuf);
+        return 0;
+    }
+
+    PRINT_DEBUG("vdso offset: %ld\n", offset);
+    ctx.addr = addr + offset;
+
+    // since vdso is a pointer
+    if (VMI_SUCCESS != vmi_read_64(vmi, &ctx, &addr))
+    {
+        PRINT_DEBUG("Failed to read vdso address\n");
+        drakvuf_release_vmi(drakvuf);
+        return 0;
+    }
+
+    PRINT_DEBUG("Got vdso address: %lx\n", addr);
+    drakvuf_release_vmi(drakvuf);
+
+    return addr;
+}
+
+/** src: https://www.geeksforgeeks.org/naive-algorithm-for-pattern-searching/ **/
+static size_t search(char* txt, char* pat, int N, int M)
```

Comment:
I understand the link and all, but please more descriptive name?

Reply:
replaced this function with `memmem`